### PR TITLE
transfer_learning_tutorial.py update from a transfer learning researcher

### PR DESCRIPTION
**Pull request by youkaichao**
Hi, awesome pytorch developers,

I'm a transfer learning researcher ([here](https://scholar.google.com/citations?user=Ke4w3KkAAAAJ&hl=en) is my research profile). I have read through the transfer learning tutorial in the [official website](https://pytorch.org/tutorials/beginner/transfer_learning_tutorial.html) and found that it is not satisfying in the following aspects:

- The dataset used is a very small dataset with only two classes, which is not practical. In fact, it makes the training very instable. 
- Hyper-parameters are not carefully tuned, and it is strange that fixing the pre-trained model should outperform fine-tuning.
- Common practices are not respected. For example, we often set a smaller learning rate for pre-trained parameters, which is violated in the tutorial.
- The performance is instable and hard to reproduce.

To overcome the above shortcomings, I have re-written the tutorial, with the following changes:

- The tiny small dataset is replaced by a subset of CUB-200-2011 with 200 classes so that the training gets stable enough.
- Random seed is set to ensure reproducibility.
- Proper tuning of hyper-parameters. Change the default learning rate to 1e-2 for newly created parameters.
- Respect common practice in transfer learning. Learning rate for pre-trained parameters is 1e-3, smaller than that of newly created parameters.

In addition, I add an advanced finetuning technique (Co-Tuning).

The logic of the script follows:
1. using a pre-trained model as fixed feature extractor (accuracy 33.29%)
2. finetune a pre-trained model (accuracy 38.44%)
3. advanced finetuning method named Co-Tuning (accuracy 41.03%)

TODO:
One thing puzzles me: how to send data zip files so that they can be hosted on the website? Like https://download.pytorch.org/tutorial/hymenoptera_data.zip

### CODE REVIEW

**Comment by youkaichao** [comment]
The dataset is now hosted [here](https://cloud.tsinghua.edu.cn/f/d7e8b74e6a4644f985f0/?dl=1).

**Comment by youkaichao** [comment]
@chsasank @soumith Please have a look when you have time :)

**Comment by chsasank** [comment]
Hey when I wrote the tutorial, I specifically chose super small dataset so that

1. People without GPUs are encouraged to solve a real problem.
2. Encourage how transfer learning can work on really small datasets.

**Comment by youkaichao** [comment]
> 1. People without GPUs are encouraged to solve a real problem.

Since PyTorch is sometimes treated as GPU-accelerated NumPy. I think the majority of PyTorch users have a GPU.

> 2. Encourage how transfer learning can work on really small datasets.

When we are dealing with a super tiny dataset, transfer learning or even deep learning may not be the appropriate solution. We may need traditional machine learning algorithms which has better explainability and about the same performance. Deep learning on a super tiny dataset brings very instable performance and bothers beginners.

In summary, in this beginners' tutorial, I think we should target the majority of PyTorch users with GPU support, and show them how to use transfer learning in a dataset with appropriate scale to apply deep learning :)

IMO, besides some minor code update to keep compatible with new PyTorch versions, the main content of this tutorial has seldom been updated since its creation. Maybe this is an opportunity for us to make the tutorial up-to-date with transfer learning research now✌️ @chsasank

**Comment by youkaichao** [comment]
CC: @holly1238 @brianjo

**Comment by brianjo** [comment]
Thanks @youkaichao! I'll look at this tomorrow.

**Comment by youkaichao** [comment]
> Thanks @youkaichao! I'll look at this tomorrow.

Thanks. I'd be happy to bring new knowledge from cutting-edge research to the pytorch community :)

**Comment by youkaichao** [comment]
Hi, @brianjo , any updates?

**Comment by brianjo** [comment]
> Hi, @brianjo , any updates?

We're still reviewing everything. I'll let you know what we come up with in the next week or so. Thanks!

**Comment by youkaichao** [comment]
Hi, @brianjo , two weeks has passed. Any updates ? :)

**Comment by youkaichao** [comment]
The Co-Tuning technique is used in this paper [Lung Sound Classification Using Co-tuning and Stochastic Normalization
](https://arxiv.org/abs/2108.01991) and showed significant improvement over vanilla fine-tuning. It would be good to update the tutorial for better guidance to beginners :)

**Comment by youkaichao** [comment]
Hi, @brianjo , two months passed away, any updates?

**Comment by youkaichao** [comment]
@brianjo  I will be happy to provide help :)
